### ISSUE TITLE
Spritewidget broken with flutter >= 1.23 

### ISSUE THREAD

**Issue by squashmode** [spritewidget/spritewidget#67]
There are some changes in the beta channel of flutter (currently Flutter 1.23.0-18.1.pre • channel beta • https://github.com/flutter/flutter.git) that break spritewidget.  This will likely impact more people as these changes make their way into stable, and therefore this should be considered a critical issue. 

1.  PointerHoverEvent has been added to flutter:
Spritewidget assumes only 4 pointer event types (Down, Move, Up, Cancel) and doesn't do anything with 'Hover'.  As a result there can be null pointer exceptions that flutter reports from the gesture handing code.  One fix is to check for null and replace with an empty list of Node in spritewidget Node::handleEvent() method.  

2. Sub-classes of PointerEvent types have been added in flutter.
In spritewidget, the 4 assumed types are recognized using their .runTimeType method in a SpriteBoxEvent.  With the addition of sub-classes of these pointer events, the runTimeTypes are not handled, and so are missed.  For example, a _TransformedPointerDownEvent, which is a PointerDownEvent is not handled, and so is missed, resulting in unexpected (or no) behavior.   A fix for this is to have the SpriteBoxEvent take, and then handle, the PointerEvents directly, rather than work with runTimeType.

Below is an example of how to fix these problems.  

```diff --git a/examples/particledesigner/lib/particle_world.dart b/examples/particledesigner/lib/particle_world.dart
index 9418ad8..837e060 100644
--- a/examples/particledesigner/lib/particle_world.dart
+++ b/examples/particledesigner/lib/particle_world.dart
@@ -31,11 +31,11 @@ class ParticleWorld extends NodeWithSize {
   }

   @override bool handleEvent(SpriteBoxEvent event) {
-    if (event.type == PointerDownEvent || event.type == PointerMoveEvent) {
+    if (event.type is PointerDownEvent || event.type is PointerMoveEvent) {
       particleSystem.insertionOffset = convertPointToNodeSpace(event.boxPosition) - const Offset(512.0, 512.0);
     }

-    if (event.type == PointerDownEvent) {
+    if (event.type is PointerDownEvent) {
       particleSystem.reset();
     }

diff --git a/lib/src/sprite_box.dart b/lib/src/sprite_box.dart
index c291dd1..3983ea9 100644
--- a/lib/src/sprite_box.dart
+++ b/lib/src/sprite_box.dart
@@ -236,12 +236,12 @@ class SpriteBox extends RenderBox {
     }

     // Pass the event down to nodes that were hit by the pointerdown
-    List<Node> targets = entry.nodeTargets;
+    List<Node> targets = entry.nodeTargets ?? <Node>[];
     for (Node node in targets) {
       // Check if this event should be dispatched
       if (node.handleMultiplePointers || event.pointer == node._handlingPointer) {
         // Dispatch event
-        bool consumedEvent = node.handleEvent(new SpriteBoxEvent(globalToLocal(event.position), event.runtimeType, event.pointer));
+        bool consumedEvent = node.handleEvent(new SpriteBoxEvent(globalToLocal(event.position), event, event.pointer));
         if (consumedEvent == null || consumedEvent)
           break;
       }
@@ -256,7 +256,7 @@ class SpriteBox extends RenderBox {

   @override
   bool hitTest(HitTestResult result, { Offset position }) {
-    result.add(new _SpriteBoxHitTestEntry(this, position));
+    result.add(_SpriteBoxHitTestEntry(this, position));
     return true;
   }
@@ -538,7 +538,7 @@ class SpriteBoxEvent {
   ///     if (event.type == PointerDownEvent) {
   ///       // Do something!
   ///     }
-  final Type type;
+  final PointerEvent type;

   /// The id of the pointer. Each pointer on the screen will have a unique pointer id.
   ///
diff --git a/lib/src/virtual_joystick.dart b/lib/src/virtual_joystick.dart
index ac1df58..df0300a 100644
--- a/lib/src/virtual_joystick.dart
+++ b/lib/src/virtual_joystick.dart
@@ -43,12 +43,12 @@ class VirtualJoystick extends NodeWithSize {

   @override
   bool handleEvent(SpriteBoxEvent event) {
-    if (event.type == PointerDownEvent) {
+    if (event.type is PointerDownEvent) {
       _pointerDownAt = event.boxPosition;
       motions.stopAll();
       _isDown = true;
     }
-    else if (event.type == PointerUpEvent || event.type == PointerCancelEvent) {
+    else if (event.type is PointerUpEvent || event.type is PointerCancelEvent) {
       _pointerDownAt = null;
       _value = Offset.zero;
       MotionTween moveToCenter = new MotionTween((a) { _handlePos = a; }, _handlePos, _center, 0.4, Curves.elasticOut);
```

**Comment by TheKashe** [comment]
Above solution doesn't work, as "is" operator doesn't work for a type class (it only works on an instance).

The only workaround I can think of is to add original event to the SpriteBoxEvent and than use "is" against the original event.

**Comment by TheKashe** [comment]
Alternative, backwards compatible solution:

```
+  
+  ///Original pointer event. Use is operator to include subtypes
+  ///     if (event.pointerEvent is PointerDownEvent) {
+  ///       // Do something!
+  ///     }
+  final PointerEvent pointerEvent;
 
   /// Creates a new SpriteBoxEvent, typically this is done internally inside the SpriteBox.
   ///
   ///     var event = new SpriteBoxEvent(new Point(50.0, 50.0), 'pointerdown', 0);
-  SpriteBoxEvent(this.boxPosition, this.type, this.pointer);
+  SpriteBoxEvent(this.boxPosition, this.pointerEvent, this.pointer);
 }

-  final Type type;
+  @Deprecated("use pointerEvent in tue future")
+  Type get type{
+    final t=pointerEvent.runtimeType;
+    if(pointerEvent is PointerDownEvent)
+      return PointerDownEvent;
+    if(pointerEvent is PointerUpEvent)
+      return PointerUpEvent;
+    if(pointerEvent is PointerMoveEvent)
+      return PointerMoveEvent;
+    if(pointerEvent is PointerCancelEvent)
+      return PointerCancelEvent;
+  }

-        bool consumedEvent = node.handleEvent(new SpriteBoxEvent(globalToLocal(event.position), event.runtimeType, event.pointer));
+        bool consumedEvent = node.handleEvent(new SpriteBoxEvent(globalToLocal(event.position), event, event.pointer));

```

**Comment by vlidholt** [comment]
Fixed in 2.0 release
